File: 프로그래머스/lv2/12909. 올바른 괄호/올바른 괄호.cpp
#include <string>
#include <iostream>
#include <stack>

using namespace std;

bool solution(string s)
{
    bool answer = true;
    stack<char> st;
    
    for (int i = 0; i < s.size(); i++) {
        if (s[i] == '(')    // 스택에 '('만 push
            st.push(s[i]);
        else if (s[i] == ')' && !st.empty())    // 스택이 비어있지 않으면 '('가 있다는 의미
            st.pop();                           // 쌍이 맞았으므로 맨 위의 '(' 제거
        else if (s[i] == ')' && st.empty()) 
            return false;
    }
    if (!st.empty())    // for문 통과 후 stack이 비어있지 않으면 '('가 남았다는 의미
            return false;
    
    return answer;
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    answer = True
    stack = []
    for i in range(len(s)):
        if s[i] == '(':
            stack.append(s[i])
        else:   # ')'일 때
            if stack:
                stack.pop()
            else:
                answer = False
                break
    if stack:
        answer = False
    return answer